### PR TITLE
Use a queue instead of array reduce for performance

### PR DESCRIPTION
**Pull request by heyheyhello**
I was suspicious of the number of function calls in gathering children so I made a performance test: [perf.link](https://perf.link/#eyJpZCI6Im03bnA0dDRmM3VxIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6Ii8vIEdlbmVyaWMgbmVzdGVkIG9iamVjdCBzdHJ1Y3R1cmVcbmNvbnN0IG1rRGF0YSA9IChuKSA9PiBbLi4uQXJyYXkobikua2V5cygpXTtcbmNvbnN0IHVwZGF0ZSA9ICgpID0%2BIHt9O1xudXBkYXRlLmlkID0gMFxudXBkYXRlLmNoaWxkcmVuID0gbWtEYXRhKDUwKS5tYXAoeCA9PiB7XG4gIGNvbnN0IGEgPSAoKSA9PiB7fTtcbiAgYS5pZCA9IHg7XG4gIGEuY2hpbGRyZW4gPSBta0RhdGEoNTApLm1hcCh5ID0%2BIHtcbiAgICBjb25zdCBiID0gKCkgPT4ge307XG4gICAgYi5pZCA9IHk7XG4gICAgYi5jaGlsZHJlbiA9IFtdO1xuICAgIHJldHVybiBiO1xuICB9KTtcbiAgcmV0dXJuIGE7XG59KTsiLCJ0ZXN0cyI6W3sibmFtZSI6IlF1ZXVlIiwiY29kZSI6ImxldCBjdXJyO1xubGV0IHdvcmsgPSAwO1xuY29uc3QgcXVldWUgPSBbXS5jb25jYXQodXBkYXRlLmNoaWxkcmVuKTtcbndoaWxlIChjdXJyID0gcXVldWUucG9wKCkpIHtcbiAgLy8gRG8gXCJ3b3JrXCJcbiAgd29yayArPSBjdXJyLmlkO1xuICBjdXJyLmNoaWxkcmVuLmZvckVhY2godSA9PiBxdWV1ZS5wdXNoKHUpKTtcbn1cbmNvbnNvbGUubG9nKHdvcmspOyIsInJ1bnMiOls3MzIsODAzLDgwMyw4OTIsODc1LDc2Nyw4MjEsMTAwMCw3MTQsOTgyLDEwMDAsODAzLDg3NSw5MTAsODIxLDg5MiwxMDcxLDkxMCw3NjcsMTAzNSw5NjQsMTE0Miw4OTIsODc1LDc1MCw4NzUsNjYwLDk4Miw3NTAsODM5LDEwMTcsMTE2MCw4MDMsMTAxNyw1MTcsODIxLDk0Niw4NTcsNjk2LDk0NiwxMTQyLDk4MiwxMDUzLDExNDIsNzY3LDgwMyw2OTYsOTEwLDc1MCw5NjQsMTAxNywxMDE3LDEwMTcsMTA1Myw2NzgsMTAxNywxMDg5LDgzOSw3NjcsNzMyLDExMjUsMTAwMCw3NjcsOTEwLDg3NSw4MzksODIxLDkyOCw4NzUsODIxLDEwODksMTAwMCw5MjgsMTA3MSw4MjEsOTI4LDk2NCw2NDIsNTcxLDYwNyw3NjcsODAzLDczMiw3MzIsOTQ2LDEwMDAsMTE2MCw3NTAsNzUwLDUxNyw4OTIsODAzLDk0Niw5ODIsMTA3MSwxMDE3LDgyMSwxMDM1LDkxMCwxMDUzXSwib3BzIjo4ODZ9LHsibmFtZSI6IkFycmF5IHJlZHVjZSIsImNvZGUiOiJjb25zdCBnZXRDaGlsZHJlbkRlZXAgPSAoY2hpbGRyZW4pID0%2BXG4gIGNoaWxkcmVuLnJlZHVjZShcbiAgICAoYWNjLCBjdXJyKSA9PiBhY2MuY29uY2F0KGN1cnIsIGdldENoaWxkcmVuRGVlcChjdXJyLmNoaWxkcmVuKSksXG4gICAgW11cbiAgKTtcbmNvbnN0IGFsbCA9IGdldENoaWxkcmVuRGVlcCh1cGRhdGUuY2hpbGRyZW4pO1xuXG5sZXQgd29yayA9IDA7XG5hbGwuZm9yRWFjaCh4ID0%2BIHtcbiAgLy8gRG8gXCJ3b3JrXCJcbiAgd29yayArPSB4LmlkO1xufSlcbmNvbnNvbGUubG9nKHdvcmspOyIsInJ1bnMiOlsxNDIsMTI1LDE3OCwxNjAsMTQyLDE2MCwxNDIsMTc4LDEyNSwxNzgsMTc4LDE0MiwxNjAsMTYwLDE2MCwxNjAsMTk2LDE5NiwxNjAsMTk2LDE5NiwyMTQsMTYwLDE5NiwxNDIsMTc4LDE5NiwxOTYsMTQyLDE3OCwxNzgsMTc4LDE2MCwxOTYsMTQyLDEwNywxNzgsMTk2LDE0MiwxOTYsMjE0LDE3OCwxNzgsMTk2LDE2MCwxMDcsMTQyLDE3OCwxNDIsMTc4LDE5NiwyMTQsMTk2LDE3OCwxMjUsMTk2LDE3OCwxNjAsMTQyLDE3OCwyMTQsMTc4LDE2MCwxNDIsMTYwLDE2MCwxNzgsMTc4LDE3OCwxNjAsMjE0LDE5NiwxOTYsMTk2LDE2MCwxNzgsMTc4LDE0MiwxMDcsMTk2LDE3OCwxNDIsMTQyLDE0MiwxOTYsMTc4LDE5NiwxNjAsMTQyLDE0MiwxNDIsMTYwLDE5NiwxOTYsMTk2LDE3OCwxNjAsMTk2LDE3OCwxOTZdLCJvcHMiOjE3MH1dLCJ1cGRhdGVkIjoiMjAyMC0xMC0xOFQwNjowOTozMS4wMTlaIn0%3D)

Firefox is 19% the speed.
image

Opera/Blink is 1% (!!!) the speed.
image

### CODE REVIEW

**Comment by luwes** [comment]
Wow this is impressive! I'll have a look at this soon. Great find!

**Comment by luwes** [comment]
checking to make sure we don't re-introduce this bug. I don't think it's covered in the tests
https://github.com/luwes/sinuous/issues/20#issuecomment-526250502

**Review by luwes — APPROVED** [review]
LGTM 👍 awesome work! thank you

**Comment by heyheyhello** [comment]
That's good to check and have tests for it. I tried to make other optimizations elsewhere and was surprised but happy that tests failed - helped me a lot in understanding how the code works.

**Comment by luwes** [comment]
still checking the tests a little, I hit a case where the tests still pass when commenting out this whole block:

I feel like there should be a test failing when we do this.

```js
    // prevChildren.forEach(u => {
    //   if (update._children.indexOf(u) === -1) {
    //     u._fresh = true;
    //   }
    // });

    // // If any children were marked as fresh remove them from the run lists.
    // let curr;
    // const queue = [].concat(update.children);
    // while (curr = queue.pop()) {
    //   if (curr._fresh) {
    //     curr._observables.forEach(o => {
    //       if (o._runObservers) o._runObservers.delete(curr);
    //     });
    //   }
    //   curr.children.forEach(u => queue.push(u));
    // }
```

**Comment by heyheyhello** [comment]
Hmm yeah. That does seem strange. As the the actual PR changes though it doesn't look like any observables could run or changes changes (to the run list for example) - I imagine deleting all children from run lists is safe regardless of the order of operations?

**Comment by luwes** [comment]
you're right, just noticed those are deleted anyway
https://github.com/luwes/sinuous/commit/700bf4c6387999fddd70a1090ea5fba68d3ea105#diff-557ac877462f43d8cd15cc29b87f96bf489abfc5ebc44fadbc958a5d10cda04e

I'm wondering if that whole block might be unneeded then because that case Ryan explained is passing its tests.

File: packages/sinuous/observable/src/observable.js [hunk 1]
```diff
@@ -172,9 +170,16 @@ function computed(observer, value) {
     });
 
     // If any children were marked as fresh remove them from the run lists.
-    const allChildren = getChildrenDeep(update._children);
-    allChildren.forEach(removeFreshChildren);
-
+    let curr;
+    const queue = [].concat(update.children);
```
Comment (luwes) [inline packages/sinuous/observable/src/observable.js:174]: just noticed, missing the underscore here. `update._children`

Comment (heyheyhello) [reply packages/sinuous/observable/src/observable.js:174]: WHOA. Huh sorry I missed that converting from a TS codebase back to JS... I'm too used to the editor catching that :(

**Pull request merged**